TechnicalDocuments/ControlTheWorld.md

```diff
@@ -0,0 +1,120 @@
+
+## Implementation
+
+Define a `struct` to provide dependencies of the environment external to the app (the "World" in which the app lives), and static configuration dependencies:
+
+```swift
+struct World {
+    var date = { Date () }
+    var calendar = Calendar.autoupdatingCurrent
+    var timezone = Timezone.autoupdatingCurrent
+    var locale = Locale.autoupdatingCurrent
+    var analytics = Analytics()
+}
+```
+
+We can make external services like `Analytics` mockable without needing protocols:
+
+```swift
+struct Analytics {
+    struct Event {
+        var name: String
+        var properties: [String: String]
+    }
+
+    var track = track(_:)
+}
+
+private func track(_ event: Analytics.Event) {
+    // track
+}
+
+extension Analytics {
+    static let mock = Analytics(track: { event in
+        print("Mock track", event)
+    })
+}
+```
+
+By pulling the implementation (the `track` function) out into a private free function, it can be substituted with a mock.
+
+## Creating the World
+
+Now all we need is to instantiate an instance of `World` called `Current` which is our current environment. By default this will instantiate the production dependencies:
+
+```swift
+var Current = World()
+```
+
+## Usage
+
+Everywhere where we'd be using a dependency, use `Current` to access it instead. For example, use `Current.date()` as opposed to `Date()`.
+
+## Testing & Simulating different environments
+
+If we want to setup a test environment we create the `World` with test values instead of the defaults:
+
+```swift
+var Current = World(
+    date: { Date(timeIntervalSinceReferenceDate: 557152051) },
+    calendar: Calendar(identifer: .gregorian),
+    timezone: Timezone(identifier: "Europe/London")!,
+    locale: Locale(identifier: "en_GB"),
+    analytics: .mock
+)
+```
+
+## Applying this to the Babylon iOS project
+
+The project uses protocols and dependency injection **heavily**. `struct` based approaches like the `Analytics` example above are probably not a realistic target for a first implementation of `Current` in our project.
+
+Instead I propose that, for now, we put everything that is currently in `AppDependencies` into `Current`, along with `date`, `calendar`, `timezone` and `locale` (which we currently are not even injecting).
+
+This will already save us a lot of DI boilerplate and give us the ability to simulate different date, calendar, timezone and locale settings easily.
+
+Once we get more familiar with using `Current`, we can explore using the `struct` approach above for analytics and AB testing services.
+
+## Where is the boundary between Current and using Dependency Injection?
+
+In previous discussions with the team it was agreed that we can / should use `Current` as far as our Builders and no further. The Builders should still continue to inject any dependencies required by View Models, View Controllers and Flow Controllers.
```

Comment:
I've used a slightly different wording. I'm not keen about "in the most specific way possible", which kind of implies that instead of injecting `VisualDependenciesProtocol` you would have to inject only the bits of it that you need (eg `brandColors` or `appColors`) separately.